fixes #44 Support apps with WatchKit

marthinfreij opened the pull request:
Projects that support WatchKit generates three build products (the iOS app, the Watch app and the WatchKit Extension) instead of normally just one.

This fix will get the iOS app product name from the build settings (PRODUCT_NAME) in the Xcode project and use that to find the right product to pass to Instruments. Added a way to set this manually in config, if needed.

marthinfreij committed:
Proper handling of projects with multiple build products, i.e. projects with WatchKit extensions

KrauseFx commented:
Thanks, looks great :+1: 

```ruby
command = "xcodebuild -#{project_key} '#{project_path}' -showBuildSettings | grep 'FULL_PRODUCT_NAME' | sed 's/[ ]*FULL_PRODUCT_NAME = //'"
```

Does this line work for every project? Maybe we should also add a way to set a custom name using the `Snapfile`, similar to the scheme. What do you think?

marthinfreij commented:
It should be safe to use for all projects, as all projects specify PRODUCT_NAME. I tried to verify this in the documentation and found that it might be better to use WRAPPER_NAME instead of FULL_PRODUCT_NAME, because the later if undocumented.

https://developer.apple.com/library/mac/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html#//apple_ref/doc/uid/TP40003931-CH3-SW1

I thought I added support for manual names by adding manual_app_name?

marthinfreij committed:
Use WRAPPER_NAME instead of FULL_PRODUCT_NAME

Use WRAPPER_NAME instead of FULL_PRODUCT_NAME as the later if
undocumented.

https://developer.apple.com/library/mac/documentation/DeveloperTools/Ref
erence/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.
html#//apple_ref/doc/uid/TP40003931-CH3-SW1

marthinfreij commented:
Changed to WRAPPER_NAME instead

KrauseFx commented on lib/snapshot/snapshot_config.rb:
```diff
@@ -140,6 +143,27 @@ def project_name
+        Helper.log.debug "Found app name: #{result}"
+      end
+
+      unless result
```
How about changing that to just `return nil` and add a `||` when it's used, see other comment.

KrauseFx commented on lib/snapshot/runner.rb:
```diff
@@ -9,7 +9,7 @@ def work(clean: true)
 
       Builder.new.build_app(clean: clean)
-      @app_path = Dir.glob("/tmp/snapshot/build/*.app").first
+      @app_path = "/tmp/snapshot/build/#{SnapshotConfig.shared_instance.app_name}"
```
```ruby
if SnapshotConfig.shared_instance.app_name
  @app_path = "/tmp/snapshot/build/#{SnapshotConfig.shared_instance.app_name}"
else
  @app_path = Dir.glob("/tmp/snapshot/build/*.app").first
end
```

KrauseFx commented:
There is no way to set it using `Snapfile` right now. But that's okay, see my comments: That way, if it can't be found, it would still fallback to the current solution.

marthinfreij committed:
Keep old way to find build product as fallback

marthinfreij commented:
Hope I got it right this time! Kind of my first time writing ruby :confused:

marthinfreij commented:
Did you see my last commit? Anything else I should change?

KrauseFx commented:
@marthinfreij thanks again for providing your pull request. I'm currently on some conferences over the next weeks, and will come back to this issue on Monday :+1:

KrauseFx commented:
Sorry, forgot to update this pull request, I already made a new release that should fix this problem: https://github.com/KrauseFx/snapshot/releases/tag/0.5.0

Let me know if that works for you :+1:

marthinfreij commented:
Ok, cool! I think your way of solving this could result in new errors in the future as you're basically guessing the path to the build product instead of asking the build configuration for the correct value.

KrauseFx commented:
What do you mean?

marthinfreij commented:
If I understand your code right, you're looping through all paths, excluding the WatchKit product paths, which will work for now. But what if Apple adds some other build product that is not a WatchKit related product? Then you have to patch this bug again. In my opinion it's safer to ask the build configuration what the name of the iOS app build target is. Then Apple could do whatever in the future.
